File: android/app/src/main/kotlin/com/automattic/wordpresstv/promo/Promo.kt
package com.automattic.wordpresstv.promo

import android.content.Context
import android.content.Intent
import android.net.Uri

/**
 * The "Code for the People" documentary promo. WordPress.tv can't host the film
 * directly, so we feature it on Home and send viewers to YouTube — opening a
 * YouTube app when the device has one, and otherwise letting them scan the QR to
 * watch on their phone. This is static, curated content (like [catalog.Catalog]);
 * it never touches the repository or data layer.
 */
object CodeForThePeople {
    const val VIDEO_ID = "8lQijrTaaGg"
    const val TITLE = "Code for the People"
    const val TAGLINE = "The human story of the open web"
    const val CREDIT = "Directed by Bao Nguyen"
    const val BLURB =
        "The open web is arguably the world's most vital invisible utility — and it's under " +
            "siege. Code for the People is a documentary short on the past, present, and contested " +
            "future of the internet: what it's for, who gets to own it, and what it takes to keep " +
            "it free. The web belongs to all of us."

    /** Encoded in the QR and opened on a phone — the short link resolves into the app. */
    const val SHARE_URL = "https://youtu.be/$VIDEO_ID"

    /** The canonical watch URL used for the on-device app hand-off. */
    const val WATCH_URL = "https://www.youtube.com/watch?v=$VIDEO_ID"

    /** Shown under the QR as human-readable text. */
    const val SHARE_LABEL = "youtu.be/$VIDEO_ID"
}

/** The YouTube apps we'd rather hand off to, most-preferred (TV) first. */
private val YOUTUBE_PACKAGES = listOf(
    "com.google.android.youtube.tv", // Android TV / Google TV YouTube
    "com.google.android.youtube", // phone/tablet YouTube (present on some devices)
)

private fun watchIntent(pkg: String? = null): Intent =
    Intent(Intent.ACTION_VIEW, Uri.parse(CodeForThePeople.WATCH_URL)).apply {
        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        if (pkg != null) setPackage(pkg)
    }

/**
 * Open the film in a YouTube app, trying the known TV/phone packages in order and
 * launching the first that's installed. Deliberately targets a concrete package
 * (never a generic `ACTION_VIEW`): a generic intent would pop the system chooser
 * and its own "no app can do this" toast when nothing handles it. Returns false
 * when no known YouTube app is present, so the caller can show our own message.
 * Relies on the `<queries>` entry in the manifest on Android 11+.
 */
fun openInYouTube(context: Context): Boolean {
    val pm = context.packageManager
    val pkg = YOUTUBE_PACKAGES.firstOrNull { watchIntent(it).resolveActivity(pm) != null } ?: return false
    return runCatching { context.startActivity(watchIntent(pkg)) }.isSuccess
}

File: apple/WordPressTV/Promo/Promo.swift
import Foundation
import UIKit

/// The "Code for the People" documentary promo. WordPress.tv can't host the film
/// directly, so we feature it on Home and send viewers to YouTube — opening a
/// YouTube app when the TV has one, and otherwise letting them scan the QR to
/// watch on their phone. This is static, curated content (like `Catalog`); it
/// never touches the repository or data layer. Mirrors the Android
/// `CodeForThePeople`.
enum CodeForThePeople {
    static let videoID = "8lQijrTaaGg"
    static let title = "Code for the People"
    static let tagline = "The human story of the open web"
    static let credit = "Directed by Bao Nguyen"
    static let blurb = """
        The open web is arguably the world's most vital invisible utility — and it's \
        under siege. Code for the People is a documentary short on the past, present, \
        and contested future of the internet: what it's for, who gets to own it, and \
        what it takes to keep it free. The web belongs to all of us.
        """

    /// Encoded in the QR and opened on a phone — the short link resolves into the app.
    static let shareURL = URL(string: "https://youtu.be/\(videoID)")!

    /// The canonical watch URL, handed to a YouTube app as a universal link.
    static let watchURL = URL(string: "https://www.youtube.com/watch?v=\(videoID)")!

    /// The YouTube app's custom-scheme deep link, tried first for a direct hand-off.
    static let appURL = URL(string: "youtube://www.youtube.com/watch?v=\(videoID)")!

    /// Shown under the QR as human-readable text.
    static let shareLabel = "youtu.be/\(videoID)"
}

/// Hands the film off to a YouTube app on the TV. tvOS has no browser fallback, so
/// when nothing can open it the caller surfaces a toast (like Android's system
/// "no app can do this"); the QR/phone path stays separately available. Mirrors the
/// Android intent-based hand-off.
@MainActor
enum YouTubeHandoff {
    /// Open the film in a YouTube app: the custom scheme first (the reliable signal
    /// on a browser-less TV), then the universal link. Reports `false` when nothing
    /// could open it, so the caller can fall back to the QR.
    static func open(_ completion: @escaping (Bool) -> Void) {
        open([CodeForThePeople.appURL, CodeForThePeople.watchURL], completion: completion)
    }

    private static func open(_ candidates: [URL], completion: @escaping (Bool) -> Void) {
        guard let url = candidates.first else {
            completion(false)
            return
        }
        UIApplication.shared.open(url, options: [:]) { success in
            if success {
                completion(true)
            } else {
                open(Array(candidates.dropFirst()), completion: completion)
            }
        }
    }
}
